pkg/pillar/cmd/monitor/ipc_server.go

```diff
@@ -0,0 +1,293 @@
+	// Ok and Err in exactly this spelling are variants or rust's Result<T, E> type
+	Ok  string `json:"Ok,omitempty"`
+	Err string `json:"Err,omitempty"`
+	Id  uint64 `json:"id"` // Id is the id of the request that this response is for
+}
+
+type IpcMessage struct {
+	Type    string          `json:"type"`
+	Message json.RawMessage `json:"message"`
+}
+
+type IPCServer struct {
+	codec *framed.ReadWriteCloser
+	// dataReady chan bool
+	ctx *monitorContext
+	sync.Mutex
+	clientConnected chan bool
+}
+
+// constructor
+func newIPCServer(ctx *monitorContext) *IPCServer {
+	return &IPCServer{
+		ctx:             ctx,
+		clientConnected: make(chan bool),
+	}
+}
+
+func (s *IPCServer) C() chan bool {
+	return s.clientConnected
+}
+
+func (s *IPCServer) HandleConnection(conn net.Conn) {
+	s.Lock()
+	defer s.Unlock()
+	// the format of the frame is length + data
+	// where the length is 16 bit unsigned integer
+	s.codec = framed.NewReadWriteCloser(conn)
+
+	go func() {
+		defer s.Close()
+		s.run()
+	}()
+
+	// notify the monitor that the client is connected
+	s.ctx.clientConnected <- true
+}
+
+// close the server
+func (s *IPCServer) Close() {
+	s.codec.Close()
+}
+
+// main loop
+func (s *IPCServer) run() {
+	// we never exit from the loop until the connection is closed
+	// other errors are logged and we continue
+	for {
+		// read request
+		req, err := s.readRequest()
+		if err != nil {
+			log.Warnf("Error reading request: %v", err)
+			// exit if EOF
+			if errors.Is(err, io.EOF) {
+				return
+			}
+			continue
+		}
+		// handle request
+		resp := s.handleRequest(req)
+		log.Noticef("Response: %v", resp)
+		// send response
+		if err := s.sendResponse(resp); err != nil {
+			if errors.Is(err, io.EOF) {
+				log.Notice("Connection closed by client")
+				return
+			}
+			log.Warnf("Error sending response: %v", err)
+		}
+	}
+}
+
+// read request
+func (s *IPCServer) readRequest() (*Request, error) {
+	frame, err := s.codec.ReadFrame()
+	if err != nil {
+		return nil, err
+	}
+	log.Noticef("Received frame: %v", string(frame))
+
+	// following code is used for debugging when #[serde(untagged)] line
+	// is commented out in the rust code in the IpcMessage struct
+	// unmarshal IpcMessage first
+	// var ipcMessage IpcMessage
+	// if err := json.Unmarshal(frame, &ipcMessage); err != nil {
+	// 	return nil, err
+	// }
+
+	var request Request
+	if err := json.Unmarshal(frame, &request); err != nil {
+		return nil, err
+	}
+	return &request, nil
+}
+
+// send response
+func (s *IPCServer) sendResponse(resp *Response) error {
+	return s.SendIpcMessage("Response", resp)
+}
+
+func (s *IPCServer) SendIpcMessage(t string, msg any) error {
+	s.Lock()
+	defer s.Unlock()
+
+	var err error
+
+	if data, err := json.Marshal(msg); err == nil {
+		ipcMessage := IpcMessage{Type: t, Message: json.RawMessage(data)}
+		if data, err = json.Marshal(ipcMessage); err == nil {
+			log.Noticef("Sending IPC message: %s", string(data))
+			_, err = s.codec.Write(data)
```

Comment:
please wrap the error so that the reader of the log can see later which function the error came from